torch/csrc/jit/passes/automatic_mixed_precision.cpp

```diff
@@ -0,0 +1,263 @@
+            use.user->s(attr::name) == "_enabled") {
+          const auto def = use.user->input(1)->node();
+          if (def->kind() != prim::Constant) {
+            AT_ERROR("Autocast argument must be a constant");
+          }
+
+          // We have an autocast instance
+          AutocastScope scope;
+          scope.instance = value;
+          scope.enabled = def->i(attr::value) ? true : false;
+          return scope;
+        }
+      }
+    }
+  }
+
+  // Not an autocast...
+  return c10::nullopt;
+}
+
+void castTensorInputs(Node* node, at::ScalarType dtype) {
+  const auto graph = node->owningGraph();
+
+  WithInsertPoint insert_point(node);
+
+  const auto dtype_value = graph->insertConstant(dtype);
+  const auto false_value = graph->insertConstant(false);
+  const auto none_value = graph->insertConstant(IValue());
+
+  std::unordered_set<Value*> casted_inputs;
+
+  for (auto input : node->inputs()) {
+    if (input->type()->kind() == TensorType::Kind) {
+      casted_inputs.insert(input);
+    }
+  }
+
+  for (auto input : casted_inputs) {
+    const auto new_input = graph->insert(
+        aten::to, {input, dtype_value, false_value, false_value, none_value});
+    node->replaceInputWith(input, new_input);
+  }
+}
+
+void castInputsToWidestType(Node* node) {
+  // Figure out the widest type
+  // (really, just looking for any float32 inputs)
+  //
+  // TODO: revisit this (do we need to consider float64 types?)
+  //
+  for (auto input : node->inputs()) {
+    if (auto tensor_type = input->type()->cast<TensorType>()) {
+      const auto dtype = tensor_type->scalarType();
+      if (!dtype.has_value() || *dtype != at::ScalarType::Half) {
+        castTensorInputs(node, at::ScalarType::Float);
+        return;
+      }
+    }
+  }
+}
+
+void handleBlock(Block* block, bool initial_state) {
+  std::stack<AutocastScope> autocast_stack;
+
+  // The current autocast enabled/disabled state
+  auto current_state = [&] {
+    return autocast_stack.empty() ? initial_state
+                                  : autocast_stack.top().enabled;
+  };
+
+  for (Node* node : block->nodes()) {
+    switch (node->kind()) {
+      case prim::CallFunction:
+      case prim::CallMethod:
+        TORCH_INTERNAL_ASSERT(false, "Calls are not expected with AMP & JIT");
+        break;
+
+      case prim::Enter:
+        if (auto autocast_scope = parseAutocast(node->input())) {
+          autocast_stack.push(*autocast_scope);
+        }
+        break;
+
+      case prim::Exit:
+        // TODO: technically we can avoid parseAutocast() here
+        if (auto autocast_scope = parseAutocast(node->input())) {
+          TORCH_INTERNAL_ASSERT(!autocast_stack.empty());
+          TORCH_INTERNAL_ASSERT(
+              autocast_stack.top().instance == autocast_scope->instance);
+          autocast_stack.pop();
+        }
+        break;
+
+      // CastPolicy::fp16 (cast all inputs to float16)
+      case aten::_convolution:
```

Comment:
@tlemo I thought we were going to put a unified list of the casted functions somewhere, and use X macros to expand it as needed in autocast_mode.cpp and wherever it needs to be used in jit.  Will that come later?  (Since it would involve changing autocast_mode.cpp heavily, maybe it's something that should be PRed against upstream first.)